Add assertions against Elements having too many neighbors

fmahebert opened the pull request:
## Proposed changes

Adds assertions to Element and Neighbor to guard against constructing with (or adding) more neighbors than `maximum_number_of_neighbors(dim)`.

The tests for Element and Neighbor are adjusted to remove cases of a 1D test creating more than one neighbor per side.

Fixes #551 

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Update Test_Element to avoid too many neighbors

fmahebert committed:
Check that Elements do not have too many neighbors

fmahebert committed:
Update Test_Neighbors to avoid too many neighbors

The 1D and 2D tests are changed to reduce the number of neighbors.
The 3D test is cleaned up a little, but no functionality changes.

nilsdeppe commented:
This looks good :) My only concern is that each commit individually builds and can pass the tests. If the assertions are added before the tests are changed it seems like this would not be true. Is there a reason not to have one commit that just fixes this? (I mean squash into a single commit)

fmahebert commented:
Whoops, I agree that the order of commits should be inverted. I kept these as separate commits because (to me) adding the assertions felt logically different from adjusting the tests... but if the consensus is to squash, I will squash.

nilsdeppe commented:
My only preference is each commit builds and passes tests, beyond that I have no preference :) :+1: thanks for fixing this :)

Edit: By "fixing this" I mean the bug in current develop and the issue :)

fmahebert commented:
Force-pushed to reorder the commits: the changes to the tests now come before the addition of the assertions. Unfortunately github's UI appears confused...

nilsdeppe reviewed (dismissed):
LGTM

I verified that the commit order is correct :+1:

fmahebert commented on src/Domain/Neighbors.cpp:
```diff
@@ -6,22 +6,31 @@
                                 OrientationMap<VolumeDim> orientation) noexcept
-    : ids_(std::move(ids)), orientation_(std::move(orientation)) {}
+    : ids_(std::move(ids)), orientation_(std::move(orientation)) {
+  ASSERT(ids_.size() <= maximum_number_of_neighbors(VolumeDim),
```
Actually, I think this is incorrect: `Neighbor` is the neighbors in just one particular direction, so the max should be a smaller number (maximum_number_of_neighbors / number_of_faces)

kidder replied:
Yes, that is correct, this is for a particular direction.  Thus the smaller number is `2^(VolumeDim - 1)` assuming a maximum 2-to-1 refinement between two adjacent Elements (which `maximum_number_of_neighbors` assumes)

fmahebert replied:
I propose renaming `maximum_number_of_neighbors` to `maximum_number_of_element_neighbors`, and adding a new similar `maximum_number_of_face_neighbors`. Thoughts?

nilsdeppe replied:
I don't know what element and face neighbors means exactly, so I would prefer using Voronoi and Neumann neighbors as the terms since these actually exist in the literature. I think in either case the documentation needs to be clear. Voronoi is all elements that share a vertex with the self element. Neumann is all elements that share part or all of a face with the self element. If we need more functions we can be explicit, but at least these names would be consistent with what the literature calls them.

fmahebert replied:
It sounds like you are distinguishing between two different definitions for the neighbors of an element, whereas I am trying to distinguish between the neighbors of an element vs. the neighbors of one face of the element.

Another suggestion, then: `maximum_number_of_neighbors` (unchanged from current) vs. `maximum_number_of_neighbors_per_direction`

fmahebert commented on src/Domain/Neighbors.cpp:
```diff
@@ -6,22 +6,31 @@
   for (const auto& id : additional_ids) {
     ids_.insert(id);
   }
+  ASSERT(ids_.size() <= maximum_number_of_neighbors(VolumeDim),
```
Same issue here.

fmahebert committed:
Add function for max number of neighbors per direction

fmahebert commented:
Updated to fix the max size of `Neighbors`, which represents the neighbors in one particular direction only, rather than all neighbors.

fmahebert committed:
Update Test_Minmod to avoid too many neighbors

Promote a 2D test with 4 neighbors to a 3D test. No change to the
test itself, but now the number of neighbors is consistent with a
maximum of 2-to-1 h-refinement.

wthrowe commented on src/Domain/MaxNumberOfNeighbors.hpp:
```diff
@@ -23,3 +23,22 @@ constexpr size_t maximum_number_of_neighbors(const size_t dim) {
+///
+/// \note Assumes a maximum 2-to-1 refinement between two adjacent Elements.
+constexpr size_t maximum_number_of_neighbors_per_direction(const size_t dim) {
+  switch (dim) {
```
Why not `two_to_the(dim - 1)`?

fmahebert replied:
How would you suggest checking for invalid `dim`? Would you omit the check?

kidder replied:
Is this function called by anything that doesn't know the dimension at compile time?

fmahebert replied:
In this PR it is only used with a compile-time-known dimension.

wthrowe replied:
I don't see why anything other than 0 should be invalid.  Handling 0 separately might be worth it, either as a `throw` or just `return 0`.  (`two_to_the((size_t)-1)` is UB.)

fmahebert replied:
I've made this change.

fmahebert committed:
Work around gcc7 ICEs in Variables

fmahebert commented:
Updated with @wthrowe's suggestion, and a new commit that (hopefully) addresses the gcc ICE in Variables.

fmahebert commented:
I've force-pushed to squash the ICE fixups, because these changes had not been reviewed.

nilsdeppe reviewed (approved):
@wthrowe @kidder could you please look at this when you get a chance?
